### PR TITLE
HPCC-16883 Add time zone support to Std.Date

### PR DESCRIPTION
**Pull request by dcamper**
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Additional tests have been added to ecllibrary/teststd/TestDate.ecl

### CODE REVIEW

**Comment by richardkchapman** [comment]
@JamesDeFabia @ghalliday Please review

**Review by JamesDeFabia — COMMENTED** [review]
Looks good from my POV. Excellent javadoc comments. @ghalliday should have the last word

**Comment by AttilaVamos** [comment]
@richardkchapman Are those teststdlibrary failures 
```
(new warnings:
+<Warning><Code>4531</Code><Filename>Date.ecl</Filename><Line>1893</Line><Source>eclcc</Source><Message>JOIN condition folded to constant, converting to an ALL join</Message></Warning>) 
```
come from your change?

**Comment by dcamper** [comment]
@richardkchapman I believe so, yes.  The new vTZInputSample attribute needs a NOFOLD to prevent constant folding.  I'll make that change.

**Comment by richardkchapman** [comment]
@ghalliday Please review/merge

**Comment by richardkchapman** [comment]
@dcamper Your latest commit seems to have been squashed so that the title only refers to the last change you made, rather than the overall impact. In order to get the changelog looking nice, please update the commit title to

HPCC-16883 Add time zone support to Std.Date 

or similar

**Comment by richardkchapman** [comment]
@dcamper Needs rebasing for a clean merge (and commit message needs fixing)

**Comment by richardkchapman** [comment]
@ghalliday Any concerns with this?

File: ecllibrary/std/Date.ecl [hunk 1]
```diff
@@ -1354,4 +1354,777 @@ EXPORT DateTime_rec CreateDateTimeFromSeconds(Seconds_t seconds) := TRANSFORM
...
+
+/**
+ * Return a list of unique location names from the hardcoded dataset.
+ * All names are in uppercase.
+ *
+ * @return              A new DATASET({STRING name}) containing the
+ *                      unique location names.
+ */
+
+EXPORT UniqueTZLocations() := FUNCTION
```
Comment (ghalliday) [inline ecllibrary/std/Date.ecl]: This should have a test case.

File: ecllibrary/teststd/Date/TestDate.ecl [hunk 2]
```diff
@@ -237,6 +241,18 @@ EXPORT TestDate := MODULE
 
     ASSERT(Date.SecondsFromDateTimeRec(vCreateDateTime) = 917872496);
 
+    ASSERT(EXISTS(Date.TimeZone.UniqueTZAbbreviations()));
+    ASSERT(EXISTS(Date.TimeZone.TZDataForLocation('NORTH AMERICA')));
```
Comment (ghalliday) [inline ecllibrary/teststd/Date/TestDate.ecl]: Better to have a slightly better test - e.g., count the number of military time zones = 25
(I never knew about half these timezones before this PR!)

File: ecllibrary/teststd/Date/TestDate.ecl [hunk 3]
```diff
@@ -237,6 +241,18 @@ EXPORT TestDate := MODULE
...
+    ASSERT(Date.TimeZone.SecondsBetweenTZ('EST', 'CST') = -3600);
+    ASSERT(Date.TimeZone.AdjustTimeTZ(120000, 'EST', 'CST') = 110000);
+    ASSERT(Date.TimeZone.ToLocalTime(120000, 'EST') = 70000);
+    ASSERT(Date.TimeZone.ToUTCTime(120000, 'EST') = 170000);
+
+    ASSERT(Date.TimeZone.AppendTZOffset(vTZInputSample, tz, seconds_to_utc)[1].seconds_to_utc_is_valid = TRUE);
+    ASSERT(Date.TimeZone.AppendTZOffset(vTZInputSample, tz, seconds_to_pt, toTimeZoneAbbrev := 'PT', toLocation := 'NORTH AMERICA')[1].seconds_to_pt_is_valid = TRUE);
+    ASSERT(Date.TimeZone.AppendTZAdjustedTime(vTZInputSample, time, tz, utc_time)[1].utc_time_is_valid = TRUE);
```
Comment (ghalliday) [inline ecllibrary/teststd/Date/TestDate.ecl]: It would also be better to check more of the results

**Review by ghalliday — COMMENTED** [review]
This looks like a really good PR.  Main comments are that it could do with a few more test cases - check all functions are tested, some more edge cases checked, and slightly more comprehensive tests.

**Comment by ghalliday** [comment]
@dcamper see comment

File: ecllibrary/std/Date.ecl [hunk 4]
```diff
@@ -1354,4 +1354,777 @@ EXPORT DateTime_rec CreateDateTimeFromSeconds(Seconds_t seconds) := TRANSFORM
...
+                        DATASET(NameRec)    names
+                    },
+                    SELF.names := DATASET(LEFT.locations, NameRec)
+                )
+        );
+
+    // Flatten collected names, so there is one name per record
+    flattenedNames := NORMALIZE
+        (
+            collectedNames(EXISTS(names)),
```
[the remark this answers is not in the document]
Comment (dcamper) [reply ecllibrary/std/Date.ecl]: Good catch.  That filter was left over from a version where not all time zones had location names.

**Comment by dcamper** [comment]
@ghalliday All comments addressed; please re-review.

Note:  There was one data change in this latest commit.  While creating tests I discovered that the TZ abbreviation 'IST' in location 'ASIA' was actually used twice for two entirely different time zones.  The original source of the tz information has that duplicate as well.  In order to differentiate those duplicate time zones, I changed the location of the two Israel-specific zones to 'ISRAEL'.  The location name is not part of a standard as far as I can tell, and this seemed to be the minimum change I could make to provide differentiation.  I have added comments regarding the data source and this deviation from that source to the file.  If you have any thoughts on this change, please let me know.

Thanks!
